breaking: drop python 3.5 support

Legorooj opened the pull request:
* drop 3.5 from the CI matrices
* bump the check in compat.py so PyInstaller no longer runs on 3.5
* Remove/update most python 3.5 references in the docs
* update requirements.txt so that 3.5-specific dependencies are removed
* update setup.cfg so that PyInstaller won't install on python 3.5
* update the readme

BoboTiG commented on tests/requirements-libraries.txt:
```diff
@@ -17,7 +17,6 @@
 # - v. 2.2 and above fails.
 Django==2.1.8  # pyup: ignore
-matplotlib==3.0.3; python_version <= '3.5'  # pyup: ignore
 matplotlib==3.2.2; python_version >= '3.6'  # pyup: ignore
```
```suggestion
matplotlib==3.2.2
```

Legorooj replied:
Missed this, thanks.

BoboTiG reviewed (commented):
Quick review. I will be able to do second round when on my computer, tomorrow.

bwoodsend reviewed (commented):
Grepping around your branch I see a couple more:
```
doc/man/pyinstaller.rst:Python 3.5—3.7, it builds smaller executables thanks to transparent
doc/index.rst:The minimum required version is now Python 3.5.
```
There's also a:
```python
@skipif(not is_py36,
        reason='Call to resolve() raises an exception in Python 3.5.')
```
in `tests/functional/test_django.py` which can now go.

The whole `class TestInvalidAsyncFunction (unittest.TestCase):` from `tests/unit/test_modulegraph/test_imports.py` is 3.5 only.

And if we wanted to go all out, get rid of all our `path = str(path)`:
```bash
$ grep -r -I -E "(\w+) = str\(\1\)"
...
tests/functional/test_apple_events.py:    tmpdir = str(tmpdir)  # Fix for Python 3.5
tests/functional/test_apple_events.py:    tmpdir = str(tmpdir)  # Fix for Python 3.5
tests/functional/test_linux_appimage.py:        appimagetool = str(appimagetool)
tests/functional/test_linux_appimage.py:    tmp_path = str(tmp_path)  # Fix for Python 3.5
```

rokm commented:
```shell
$ pss is_py35
./PyInstaller/compat.py
33:is_py35 = sys.version_info >= (3, 5)
```
We probably don't need `is_py35` anymore? Unless we're keeping it to avoid breaking any external users?

```shell
$ pss is_py36
./tests/functional/test_linux_appimage.py
24:from PyInstaller.compat import is_py36
32:    if not is_py36:

./tests/functional/test_django.py
19:from PyInstaller.compat import is_py36
31:@skipif(not is_py36,

./doc/hooks.rst
253:``is_py36``, ``is_py37``, ``is_py38``, ``is_py39``:

./PyInstaller/compat.py
34:is_py36 = sys.version_info >= (3, 6)
```

Since `is_py36` is now always `True`, we probably don't need it, either. Or at least the cases where it is used should be simplified...


And some more references to "3.5" - not all of them are to be removed, e.g. in change logs and comments outlining the historical evolution of the code, but I'm leaving the list as-is for completeness sake:
```
$ pss "3\.5"
./.travis.yml
8:# appears on Trusty (Ubuntu 14.04) when testing Linux Python 3.5 and 3.6 with

./tests/functional/test_apple_events.py
34:    tmpdir = str(tmpdir)  # Fix for Python 3.5
65:    tmpdir = str(tmpdir)  # Fix for Python 3.5

./tests/functional/test_linux_appimage.py
42:    tmp_path = str(tmp_path)  # Fix for Python 3.5

./tests/functional/test_django.py
27:# the call to ``resolve()`` causes Python 3.5 to raise an exception that
32:        reason='Call to resolve() raises an exception in Python 3.5.')

./tests/functional/scripts/pyi_getfilesystemencoding.py
20:# On Windows it should be still mbcs up to Python 3.5

./tests/unit/test_modulegraph/test_imports.py
537:    @unittest.skipUnless(sys.version_info[:2] == (3,5), "Requires python 3.5")
539:        # In python 3.5 the following function is invalid:

./news/5439.core.rst
1:Drop support for python 3.5; EOL since September 2020.
./doc/CREDITS.rst
56:Contributions to PyInstaller 3.5
294:- Andrei Fokau - Python 3.5 fixes.

./doc/CHANGES-3.rst
281:* (Windows) For Python >= 3.5 targeting *Windows < 10*, the developer needs to
294:* Fix ImportError for sysconfig for 3.5.4 Conda (#3105, #3106).
584:* (win32) Add ``vcruntime140.dll`` to_win_includes for Python 3.5 an 3.6
677:* (Windows) For Python >= 3.5 targeting *Windows < 10*, the developer needs to
763:- Apps built with Windows 10 and Python 3.5 may not run on Windows versions
835:- Apps built with Windows 10 and Python 3.5 may not run on Windows versions
845:- Python 3 support (3.3 / 3.4 / 3.5).
914:- Apps built with Windows 10 and Python 3.5 may not run on Windows versions

./doc/index.rst
48:The minimum required version is now Python 3.5.

./doc/CHANGES.rst
246:  3.5. The last version supporting Python 2.7 was PyInstaller 3.6.
508:3.5 (2019-07-09)

./doc/development/venv.rst
37:    env PYTHON_CONFIGURE_OPTS="--enable-shared" pyenv install 3.5.0
39:* setup virtualenv ``pyenv virtualenv 3.5.0 venvname``

./doc/man/pyinstaller.rst
26:Python 3.5—3.7, it builds smaller executables thanks to transparent

./PyInstaller/hooks/hook-shelve.py
11:# Tested on Windows 7 x64 With Pyhton 3.5

./PyInstaller/lib/modulegraph/modulegraph.py
2144:                        # In Python 3.5 some syntax problems with async

./PyInstaller/depend/dylib.py
94:    # Python 3.5 uses the Univeral C Runtime which consists of these DLLs:

./PyInstaller/depend/bindepend.py
341:    FIXME: Can this be removed since we now only support Python 3.5+?

./PyInstaller/building/build_main.py
495:                # In python 3.5 (PEP-488) .pyo files were replaced by
```

BoboTiG commented:
@rokm out of curiosity, is `pss` a custom alias of yours?

rokm commented:
> @rokm out of curiosity, is `pss` a custom alias of yours?

No, it's a progam: https://github.com/eliben/pss

Legorooj commented:
@bwoodsend

 I missed the one in `doc/index.rst`, must have not actually done it when I went into the file to change it and the others in there. 

I'm pretty sure I don't actually need to manually edit anything inside the `doc/man` folder, because iirc they're autogenerated. However I don't actually know which files are autogenerated, so I've added it because I can't see any harm. Thanks for pointing it out so I reconsidered it.

Removed the `skipif(not is_py36)`. 

> The whole class `TestInvalidAsyncFunction (unittest.TestCase):` from `tests/unit/test_modulegraph/test_imports.py` is 3.5 only.L

I'm leaving modulegraph entirely alone. Will be replacing it for 4.3.

> And if we wanted to go all out, get rid of all our `path = str(path)`:

Removed most of the instances that appear to be python 3.5 specific.

Legorooj commented:
> We probably don't need `is_py35` anymore? Unless we're keeping it to avoid breaking any external users?
Correct.

> Since is_py36 is now always True, we probably don't need it, either. Or at least the cases where it is used should be simplified...

Legorooj commented:
> We probably don't need `is_py35` anymore? Unless we're keeping it to avoid breaking any external users?

Correct.

> Since is_py36 is now always True, we probably don't need it, either. Or at least the cases where it is used should be simplified...

Leaving this as it is; we can replace it when there are other changes in/near it's use-cases.

After reviewing the other references I don't think any changes are required there.

Legorooj commented on doc/usage.rst:
```diff
@@ -464,9 +464,9 @@ You must make a unique version of the app for each word-length supported.
+For **Python >= 3.6** targeting *Windows < 10*, the developer needs to take
 special care to include the Visual C++ run-time .dlls:
-Python 3.5 uses Visual Studio 2015 run-time, which has been renamed into
+Python 3.6 uses Visual Studio 2015 run-time, which has been renamed into
```
Edited.

BoboTiG commented on doc/man/pyinstaller.rst:
```diff
@@ -23,7 +23,7 @@ PyInstaller is a program that freezes (packages) Python programs into
 FreeBSD, OpenBSD, Solaris and AIX.
 Its main advantages over similar tools are that PyInstaller works with
-Python 3.5—3.7, it builds smaller executables thanks to transparent
+Python 3.6—3.7, it builds smaller executables thanks to transparent
```
Not related to that PR, but this should be:
```suggestion
Python 3.6—3.9, it builds smaller executables thanks to transparent
```

Legorooj replied:
Ah thanks, somehow didn't spot this.

BoboTiG commented on tests/functional/test_linux_appimage.py:
```diff
@@ -21,25 +21,19 @@
-        os.chmod(appimagetool, st.st_mode | stat.S_IXUSR)
-
-    tmp_path = str(tmp_path)  # Fix for Python 3.5
+    if not appimagetool.touch():
```
It [seems](https://github.com/python/cpython/blob/79d2e62c008446fbbc6f264bb8a30e2d38b6ff58/Lib/pathlib.py#L1289-L1308) this will work only if you pass `exist_ok=False`. But I am not sure of that change, is it a better way to check for execution rights?

Legorooj replied:
Hmm, I'm not sure now. I'll update it to `os.access(appimagetool)`.
